### PR TITLE
mirror-nixos-branch: create a "deployment" on GitHub on new releases

### PR DESCRIPTION
**Pull request by grahamc**
The environment is the channel name, an example of deployments can be
found at https://api.github.com/repos/grahamc/garbage/deployments

Specify GITHUB_OAUTH_TOKEN with a GitHub Personal Access Token which
has ONLY repo_deployment permissions.

Closes https://github.com/NixOS/nixos-channel-scripts/issues/10

We'd want this code to run once per new revision, which is why I put it where I did.

Configuration updated here: https://github.com/NixOS/nixos-org-configurations/pull/24

You'd need to add that .github/token file.

### CODE REVIEW

**Comment by grahamc** [comment]
cc @edolstra @domenkozar

**Comment by grahamc** [comment]
Btw I tested this by:

1. Deleting all the other code around it
2. Hardcoding rev and channelName
3. Running nix-build on a patched default.nix, where I deleted the custom
nix / complications
4. Running the result/bin/script

It did work and it should work but I haven't tested it in context, I'm not
sure how I could.

**Comment by domenkozar** [comment]
I like this.

**Comment by grahamc** [comment]
bump @edolstra

**Comment by domenkozar** [comment]
@edolstra this would be lovely :)

**Comment by edolstra** [comment]
I don't know anything about Github deployments. What does this do exactly?

**Comment by grahamc** [comment]
For nixpkgs, people can query the list of deployments for the repository and know each time each channel was updated. It is just for record keeping.

The deployments URL would look like this:

```json
[
  {
    "url": "https://api.github.com/repos/nixos/nixpkgs/deployments/...",
    "id": 28824381,
    "sha": "7d440a94b541d7cd3f2fb51c2e8a8d085c732d55",
    "ref": "7d440a94b541d7cd3f2fb51c2e8a8d085c732d55",
    "task": "deploy",
    "payload": {

    },
    "environment": "release-16.09",
    "description": null,
    "creator": {
      "login": "grahamc",
      "id": 76716,
      "avatar_url": "https://avatars0.githubusercontent.com/u/76716?v=3",
      "gravatar_id": "",
      "url": "https://api.github.com/users/grahamc",
      "html_url": "https://github.com/grahamc",
      "followers_url": "https://api.github.com/users/grahamc/followers",
      "following_url": "https://api.github.com/users/grahamc/following{/other_user}",
      "gists_url": "https://api.github.com/users/grahamc/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/grahamc/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/grahamc/subscriptions",
      "organizations_url": "https://api.github.com/users/grahamc/orgs",
      "repos_url": "https://api.github.com/users/grahamc/repos",
      "events_url": "https://api.github.com/users/grahamc/events{/privacy}",
      "received_events_url": "https://api.github.com/users/grahamc/received_events",
      "type": "User",
      "site_admin": false
    },
    "created_at": "2017-03-06T00:40:09Z",
    "updated_at": "2017-03-06T00:40:09Z",
    "statuses_url": "https://api.github.com/repos/nixos/nixpkgs/deployments/.../statuses",
    "repository_url": "https://api.github.com/repos/nixos/nixpkgs"
  },
]
```

A deployment on GitHub _can_ be used to trigger actual actions on a server (via webhooks), but in this case is just used as a marker. No actions will take place: in order for actual actions to run we would need to add more webhooks to the repository, and set up daemons on the servers.

**Comment by domenkozar** [comment]
So this doesn't point to the cloudfront url for the channel update. This would then only be useful to list channel updates on github and parse all the channel information.

File: mirror-nixos-branch.pl [hunk 1]
```diff
@@ -89,9 +90,41 @@ sub fetch {
...
+                                      'required_contexts' => [],
+                                      'auto_merge' => \0,
+                                      'environment' => $channel_name
+                                  }
+        );
+    github_post($deploy_resp->{statuses_url},
+                {
+                    'state' => 'success'
+                }
+        );
```
Comment (edolstra) [inline mirror-nixos-branch.pl]: Is it intentional that the deployment is created *before* the release is uploaded to Cloudfront?

What happens if the script gets interrupted? Will the next run create a duplicate deployment?

**Pull request closed without merging**